File: js/test/frame.test.js
// Frame classification tests. Verdicts mirror the shared corpus
// (spec/corpus/framing.json, protocol.md §2). Byte-for-byte agreement with the
// other implementations over that corpus is checked by bin/interop_checks.js.
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { classify, encode, HANDSHAKE_CAP, TRANSPORT_CAP } from '../src/frame.js';

const buf = (s) => Buffer.from(s, 'utf8');

test('accepts a simple object', () => {
  const { obj, reason } = classify(buf('{"t":"bmx1","v":3}\n'), HANDSHAKE_CAP);
  assert.equal(reason, undefined);
  assert.equal(obj.t, 'bmx1');
});

test('verdicts match the corpus shapes', () => {
  const cases = [
    ['no-newline', '{"a":1}', TRANSPORT_CAP, 'no-newline'],
    ['empty', '\n', TRANSPORT_CAP, 'empty'],
    ['invalid-json (garbage)', 'not json\n', TRANSPORT_CAP, 'invalid-json'],
    ['invalid-json (interior newline)', '{"a":\n1}\n', TRANSPORT_CAP, 'invalid-json'],
    ['trailing-data', '{"a":1} X\n', TRANSPORT_CAP, 'trailing-data'],
    ['not-an-object', '[1,2,3]\n', TRANSPORT_CAP, 'not-an-object'],
    ['oversize', '{"a":1}\n', 4, 'oversize'],
  ];
  for (const [name, raw, cap, want] of cases) {
    assert.equal(classify(buf(raw), cap).reason, want, name);
  }
});

test('rejects invalid UTF-8', () => {
  const raw = Buffer.concat([Buffer.from('{"a":"'), Buffer.from([0xff, 0xfe]), Buffer.from('"}\n')]);
  assert.equal(classify(raw, HANDSHAKE_CAP).reason, 'invalid-utf8');
});

test('nested object with braces inside strings scans correctly', () => {
  const { obj, reason } = classify(buf('{"a":{"b":"}]"},"c":[1,{"d":2}]}\n'), TRANSPORT_CAP);
  assert.equal(reason, undefined);
  assert.deepEqual(obj.c, [1, { d: 2 }]);
});

test('encode round-trips through classify', () => {
  const enc = encode({ t: 'bmx1', mesh: 'acme' });
  assert.equal(enc[enc.length - 1], 0x0a);
  const { obj } = classify(enc, HANDSHAKE_CAP);
  assert.equal(obj.mesh, 'acme');
});

File: elixir/test/frame_test.exs
defmodule Bonemesh.FrameTest do
  use ExUnit.Case, async: true

  alias Bonemesh.Frame

  # Cases mirror the shared corpus (spec/corpus/framing.json). Byte-for-byte
  # agreement with the Java and Go classifiers over that corpus is checked by
  # interop/check-framing-elixir.sh.

  defp accept?(raw), do: match?({:ok, _}, Frame.classify(raw, Frame.transport_cap()))

  defp reject(raw), do: Frame.classify(raw, Frame.transport_cap())

  test "simple object accepts", do: assert(accept?(~s({"a":1}) <> "\n"))

  test "missing trailing newline", do: assert(reject(~s({"a":1})) == {:reject, "no-newline"})

  test "empty line", do: assert(reject("\n") == {:reject, "empty"})

  test "not json", do: assert(reject("not json\n") == {:reject, "invalid-json"})

  test "json array is not an object",
    do: assert(reject("[1,2,3]\n") == {:reject, "not-an-object"})

  test "interior newline splits frame",
    do: assert(reject(~s({"a":) <> "\n1}\n") == {:reject, "invalid-json"})

  test "invalid utf8" do
    raw = <<?{, ?", ?a, ?", ?:, ?", 0xFF, ?", ?}, ?\n>>
    assert reject(raw) == {:reject, "invalid-utf8"}
  end

  test "trailing garbage after object",
    do: assert(reject(~s({"a":1} X\n)) == {:reject, "trailing-data"})

  test "frame exactly at the cap accepts" do
    assert accept?(line_of_length(Frame.transport_cap()))
  end

  test "frame one byte over the cap is oversize" do
    assert Frame.classify(line_of_length(Frame.transport_cap() + 1), Frame.transport_cap()) ==
             {:reject, "oversize"}
  end

  defp line_of_length(n) do
    fill = n - byte_size(~s({"p":""})) - 1
    ~s({"p":") <> String.duplicate("A", fill) <> ~s("}) <> "\n"
  end
end
